Add wildcard/multiple server support to nginx_conf resource

jerryaldrichiii opened the pull request:
Fixes #2035 
Fixes #1998

jerryaldrichiii committed:
Add wildcard/multiple server support to nginx_conf

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii commented on lib/resources/nginx_conf.rb:
```diff
@@ -93,8 +96,9 @@ def resolve_references(data, rel_path)
+            .map { |x| find_files(x) }.flatten
             .map { |path| parse_nginx(path) }
-            .map { |e| data.merge!(e) }
+            .map { |e| data.merge!(e) { |_, v1, v2| v1 + v2 } }
```
I feel like this is wrong...just not sure how to make it better. When loading server configs from other files (e.g. `/etc/nginx/conf.d/*.conf`) the `merge!` would clobber the `server` value and replace it. This appends the list to the key...but that assumes that `v1` and `v2` will always be lists...I don't like that assumption.

arlimus replied:
should probably be an `each`; also let's extract this piece as it feels to cover some complexity (for deep merging)

adamleff replied:
By "lists" to you mean `Array`s?  Why wouldn't they be? If that's a concern, how do we protect against it?

We own the entire codepath here, so if there's additional promises that need to be made, we can make them.

adamleff commented on test/unit/mock/cmd/find-nginx-confd-example-conf:
```diff
@@ -0,0 +1 @@
+/etc/nginx/conf.d/example.conf
```
For completeness of testing, I'd change this to return multiple files to ensure the code actually iterates over each of them.

adamleff reviewed (changes requested):
This looks like a great first start.

I left a note about adding additional test coverage by having it find multiple files. We should make sure when it counters multiple files that it properly iterates of each.

We also need to expand the test coverage to ensure that data from each file is present. So if your sample files include multiple servers, we should have a test that is able to pluck unique data out of each server to ensure we're properly including all data and not accidentally clobbering another server record. That should help raise the confidence level of this change.

arlimus committed:
separate the merge function for maps in nginx_conf

Signed-off-by: Dominik Richter <dominik.richter@gmail.com>

arlimus commented on lib/resources/nginx_conf.rb:
```diff
@@ -93,13 +96,29 @@ def resolve_references(data, rel_path)
       Hash[data.map { |k, v| [k, resolve_references(v, rel_path)] }]
     end
+
+    # Merge conf into data list of parameters.
```
Needs better docs string 😁

arlimus reviewed (commented):
Agree with Adam's multi-file test case.
Otherwise this looks fantastic, thank you so much!!

jerryaldrichiii committed:
Address PR feedback

  - Add multiple mocked files in `/etc/nginx/conf.d`
  - Add multiple server definitions in file selected by wildcard
  - Modify comments in @Dominik's addition

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

jerryaldrichiii committed:
Add documentation fields

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

arlimus reviewed (approved):
Thank you for the great updates!!

adamleff commented on lib/resources/nginx_conf.rb:
```diff
@@ -93,13 +96,33 @@ def resolve_references(data, rel_path)
+      # Step through all conf items and create combined return value
+      data.merge!(conf) do |_, v1, v2|
+        # If both the data field and the conf field are arrays, then combine them
+        next v1 + v2 if v1.is_a?(Array) && v2.is_a?(Array)
```
Maybe it's just me, but this use of `next` feels pretty awkward. Any chance I can convince you guys to make this an `if ... elsif ... else` instead? I think it would read so much cleaner.

jerryaldrichiii committed:
Address more feedback

  - Make tests more descriptive by checking parsed content
  - Refactor `next` logic to use if/elsif/else

Signed-off-by: Jerry Aldrich <jerryaldrichiii@gmail.com>

rx294 commented on lib/resources/nginx_conf.rb:
```diff
@@ -93,13 +96,37 @@ def resolve_references(data, rel_path)
       if data.key?('include')
         data.delete('include').flatten
             .map { |x| File.expand_path(x, rel_path) }
+            .map { |x| find_files(x) }.flatten
```
Thank you for the fix Jerry. 
However this solution does not always include files (included with '*') because of issue#2135 (find_files uses the command resource).
I just tested this code over an ssh connection and it does not parse all the included conf files;
It works fine when I tested by changing the find_files code to use bash resource instead of command resource.
Hope you can find a solution for it.

Thank you again Jerry,
Rony

adamleff replied:
I know we have a logged issue about the command resource specifically, but we should log an issue for the `find_files` method too so we don't lose that context, and link it to this PR. I'm not going to hold off merging this PR since it's still a step in the right direction and will work fine for users not executing via sudo.

jerryaldrichiii replied:
https://giphy.com/gifs/reaction-LSmULmByAQHQs/fullscreen

rx294 replied:
Haha sorry @jerryaldrichiii . 
Thank you @adamleff I have posted and issue for find_files. issue #2157
